### PR TITLE
Velocity support

### PR DESCRIPTION
**Pull request by juanmuscaria**
### Summary of your change
This pull request adds support for Velocity.
It's currently based on the bungeecord version with all the logic replaced with Velocity-equivalent classes. It's my first time working with bungeecord and velocity so please correct any mistake I've made. I removed all floodgate/hooks related code because I'm unable to test bedrock connections. 
It also needs some cleanups, I made it in a rush due to needing to migrate to velocity asap. Tomorrow I'll cleanup the code and do more testings  together with implementing requested changes.
In my current testings no issues raised so far and it's working with an authme fork for velocity
image

### Related issue
Related to #549

### CODE REVIEW

File: velocity/pom.xml
Comment (games647) [inline velocity/pom.xml:94]: Please remove not needed repositories. It reduces build time, because it scans those repositories for dependencies too.

So only velocity-repo should be needed.

File: velocity/pom.xml
Comment (games647) [inline velocity/pom.xml:129]: SQLite doesn't provide the same performance than MySQL/MariaDB instance. Therefore we don't ship it for proxy instances that likely have a higher user base.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:39]: Please add the other meta data too. The version number could be quite useful for example in server logs and more. You could let them generated from the Maven project data like I did in ChangeSkin.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:49]: `java.util.Logger` is only used for backwards compatibility. If possible, we can use SL4F directly, not need to create a wrapper.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:75]: A constant like mentioned above for the annotation.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/task/ForceLoginTask.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/task/ForceLoginTask.java:134]: This could end up wrong. If there is another login request is coming in, this would evaluate to true even though it's not the same connection or even person.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/task/ForceLoginTask.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/task/ForceLoginTask.java:101]: Always use the logger for the exceptions, to ensure it has the correct output.

**Review by games647 — CHANGES_REQUESTED** [review]
Thanks for working on it. I took a quick peak at it. I'll take a deeper look later when I have some spare time.

The proxy ID system might not be necessary, because Velocity doesn't seem to suffer from malicious player plugin messages. This could be the case for players directly connecting to the server. However there is still the risk that a player could send a force login plugin message. If the proxy login crashed or is not enabled (including by mistake), the message could get forwarded to the Spigot server, so we need to verify if it's still necessary for this case.

Another thing is the asynchronous event processing. This is a main feature of velocity and is extremely important for anti bot protection. I didn't spend much time with it, but it seems to fix the design flaw BungeeCord has. I think it's important that we integrate it right from the beginning.

https://velocitypowered.com/wiki/developers/event-api/#:~:text=Handling%20events%20asynchronously

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (juanmuscaria) [reply velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:49]: Velocity only provides you with `java.util.logging.Logger` for logging, so I did the conversion like it's made on the bukkit and bungee version. I avoided messing around anything outside the velocity implementation to prevent making any unwanted change.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/task/ForceLoginTask.java
Comment (juanmuscaria) [reply velocity/src/main/java/com/github/games647/fastlogin/velocity/task/ForceLoginTask.java:134]: I missed that there was an equivalent for `#isConnected()`, my bad.

**Comment by juanmuscaria** [comment]
A few other things I'll have to look into will be testing each individual option and implement what I missed. I'll not be able to implement anything related to floodgate due to me being unable to test the implementation, but a velocity version of floodgate is available to integrate compatibility .

File: velocity/pom.xml
Comment (games647) [reply velocity/pom.xml:119]: Thanks, but then we should also relocate the package to prevent any version breakages.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (games647) [reply velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:75]: I thought Velocity provides out of the box support for it, because the guide also indicates it: https://velocitypowered.com/wiki/developers/api-basics/

**Comment by games647** [comment]
>Adding a new configuration option for a "secret" could work as replacement for the proxy id, other option is using velocity's built-in proxy secret and generate an uuid from it. Velocity allows the legacy bungee way to connect to a server, and probably most people will still use it because of older server version it it's probably a good idea to have something similar to proxy id.

For this case, we still need it. Otherwise it could cause security issues with an invalid setup. In theory the current implementation of FastLogin is not really secure. A better solution would be the implementation of Velocity with a correct signature like HMAC, but that doesn't protect against replay attacks. However it seems to be necessary in our case.

For this PR, I think it's the best to port the existing implementation and save a random UUID in the plugins folder of the proxy. So we could use that as a persistent ID. Any breaking changes like above (signing changes), should be handled in a seperate PR.

>The listeners I could probably remove the usage of the async scheduler because the event is already async and makes use of continuation.

That's kind of what I meant. Asynchronous processing of the event, so multiple users could process the event in parallel. As far as I know that's what the continuation feature promises. However I noticed now that you are already integrated it for the relevant part. The other events don't need it.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (juanmuscaria) [reply velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:75]: My bad, it indeed supports it, the error it was generating was not helpful but I finally tracked down to the issue being the relocation of org.slf4j.Logger on the shade plugin which caused the inject to fail.

File: velocity/pom.xml
Comment (juanmuscaria) [reply velocity/pom.xml:119]: I don't think it's possible to relocate an jdbc driver without also changing a few things in de driver itself, and seems like it's normal for a velocity plugin to ship the needed database driver. In case the multiple drivers are present only the first driver loaded will be used, so the end user can either overwrite the driver if needed my manually adding it to the classpath.

**Comment by juanmuscaria** [comment]
I did some testings and I think the velocity version is already in pair with the bungee version and ready for use, the only thing that will need to be implemented by someone else is floodgate support together with hooks for velocity auth plugins since so far I don't know any auth plugin for velocity.

File: velocity/pom.xml
Comment (juanmuscaria) [reply velocity/pom.xml:124]: With the way FastLogin loads drivers it can be relocated, users can use `fastlogin.mysql.cj.jdbc.Driver` on the configuration file to load the mysql driver or use any other driver provided by other plugins or from the classpath itself,

**Comment by games647** [comment]
Do you tried the MySQL driver change? You could be correct that it doesn't work.

**Comment by juanmuscaria** [comment]
Using the following config
```
driver: 'fastlogin.mysql.cj.jdbc.Driver'
host: '127.0.0.1'
port: 3306
database: 'test'
username: 'test'
password: 'password'
```
Seems to work fine with the shaded mysql driver, I know the sqlite driver will have problems because it uses native, the plugin does fail to load if no driver is shaded

**Review by games647 — APPROVED** [review]

File: velocity/pom.xml
Comment (games647) [inline velocity/pom.xml:83]: Please mentioned why we are excluding slf4j, like it's included in Velocity. It's better to be explicit for later times.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:73]: Please make this a global constant (`static`). The compiler can then inline it, not really relevant for performance, but a good convention.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:167]: This should be removed.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/listener/ConnectListener.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/listener/ConnectListener.java:98]: Good point, only texture ones if that is possible. This should make it future proof.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/listener/ConnectListener.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/listener/ConnectListener.java:127]: Correct me if I am wrong, but asserts doesn't work in without special runtime parameters. This is better solved in other programming languages ;)

**Review by games647 — CHANGES_REQUESTED** [review]
Looks good, only some minor things. However I need some more time to check for security issues before merging it just to be sure.

**Comment by games647** [comment]
> Using the following config
> 
> ```
> driver: 'fastlogin.mysql.cj.jdbc.Driver'
> host: '127.0.0.1'
> port: 3306
> database: 'test'
> username: 'test'
> password: 'password'
> ```
> 
> Seems to work fine with the shaded mysql driver, I know the sqlite driver will have problems because it uses native, the plugin does fail to load if no driver is shaded

Good to hear, thanks.

**Comment by games647** [comment]
> Using the following config
> 
> ```
> driver: 'fastlogin.mysql.cj.jdbc.Driver'
> host: '127.0.0.1'
> port: 3306
> database: 'test'
> username: 'test'
> password: 'password'
> ```
> 
> Seems to work fine with the shaded mysql driver, I know the sqlite driver will have problems because it uses native, the plugin does fail to load if no driver is shaded

We should mention this maybe. Could you update the configuration file to include a note for the different driver name, but good to hear that it works.

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (games647) [inline velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:89]: Please add a close call for the core component. It will make sure that we are safely closing our resources, mainly the Hikari pool and our own scheduler that runs tasks velocity is not aware of.

**Review by games647 — CHANGES_REQUESTED** [review]

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java
Comment (games647) [reply velocity/src/main/java/com/github/games647/fastlogin/velocity/FastLoginVelocity.java:89]: Example:

https://github.com/games647/FastLogin/blob/ec7c421f839df286ceae52d1fb89d48556972e3a/bungee/src/main/java/com/github/games647/fastlogin/bungee/FastLoginBungee.java#L103

File: velocity/src/main/java/com/github/games647/fastlogin/velocity/listener/ConnectListener.java
Comment (juanmuscaria) [reply velocity/src/main/java/com/github/games647/fastlogin/velocity/listener/ConnectListener.java:127]: That was a left over from when I was testing for a suitable object for the player session. I'll be removing it.

**Review by games647 — APPROVED** [review]

**Pull request merged**
